layers/ntp.go

```diff
@@ -0,0 +1,450 @@
+
+	// If the data block is too short to be a NTP record, then return an error.
+	//
+	if len(data) < ntp_minimum_record_size_in_bytes {
+		df.SetTruncated()
+		return fmt.Errorf("NTP packet too short")
+	}
+
+	// RFC 5905 does not appear to define a maximum NTP record length.
+	// The protocol allows "extension fields" to be included in the record,
+	// and states about these fields:"
+	//
+	//     "While the minimum field length containing required fields is
+	//      four words (16 octets), a maximum field length remains to be
+	//      established."
+	//
+	// For this reason, the packet length is not checked here for being too long.
+
+	// NTP type embeds type BaseLayer which contains two fields:
+	//    Contents is supposed to contain the bytes of the data at this level.
+	//    Payload is supposed to contain the payload of this level.
+	// Here we set the baselayer to be the bytes of the NTP record.
+	//
+	d.BaseLayer = BaseLayer{Contents: data[:len(data)]}
+
+	// Extract the fields from the block of bytes.
+	// To make sense of this, refer to the packet diagram
+	// above and the section on endian conventions.
+
+	// The first few fields are all packed into the first 32 bits. Unpack them.
+	f := binary.BigEndian.Uint32(data[0:4])
+	d.LeapIndicator = uint8((f & 0xC0000000) >> 30)
+	d.Version = uint8((f & 0x38000000) >> 27)
+	d.Mode = uint8((f & 0x07000000) >> 24)
+	d.Stratum = uint8((f & 0x00FF0000) >> 16)
+	d.Poll = int8((f & 0x0000FF00) >> 8)
+	d.Precision = int8((f & 0x000000FF) >> 0)
+
+	// The remaining fields can just be copied in big endian order.
+	d.RootDelay = binary.BigEndian.Uint32(data[4:8])
+	d.RootDispersion = binary.BigEndian.Uint32(data[8:12])
+	d.ReferenceID = binary.BigEndian.Uint32(data[12:16])
+	d.ReferenceTimestamp = binary.BigEndian.Uint64(data[16:24])
+	d.OriginTimestamp = binary.BigEndian.Uint64(data[24:32])
+	d.ReceiveTimestamp = binary.BigEndian.Uint64(data[32:40])
+	d.TransmitTimestamp = binary.BigEndian.Uint64(data[40:48])
+
+	// This layer does not attempt to analyse the extension bytes.
+	// But if there are any, we'd like the user to know. So we just
+	// place them all in an ExtensionBytes field.
+	d.ExtensionBytes = data[48:]
+
+	// Return no error.
+	return nil
+}
+
+//******************************************************************************
+
+// CanDecode() returns a set of layers that NTP objects can decode.
+// As NTP objects can only decide the NTP layer, we can return just that layer.
+// Apparently a single layer type implements LayerClass.
+//
+func (d *NTP) CanDecode() gopacket.LayerClass {
+	return LayerTypeNTP
+}
+
+//******************************************************************************
+
+// NextLayerType() specifies the next layer that GoPacket should attempt to
+// analyse after this (NTP) layer. As NTP packets do not contain any payload
+// bytes, there are no further layers to analyse.
+//
+// FIX: What is the GoPacket convention here for indicating that no
```

Comment:
Use LayerTypeZero, for example see https://github.com/google/gopacket/blob/master/base.go#L43